source/common/router/scoped_rds.cc

```diff
@@ -101,70 +107,168 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+                const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+            thread_local_scoped_config->removeRoutingScope(scope_name);
+            return config;
+          },
+          [to_be_deleted]() { /*to_be_deleted will be destructed in main thread.*/ });
+      any_applied = true;
+      ENVOY_LOG(debug, "srds: remove scoped route '{}'", scope_name);
+    }
   }
-
-  for (const auto& scoped_route : scoped_routes_to_remove) {
-    const std::string scoped_route_name = scoped_route.first;
-    ENVOY_LOG(debug, "srds: remove scoped route '{}'", scoped_route_name);
-    scoped_config_manager_.removeRoutingScope(scoped_route_name);
-    applyConfigUpdate([scoped_route_name](const ConfigProvider::ConfigConstSharedPtr& config)
-                          -> ConfigProvider::ConfigConstSharedPtr {
-      // In place update.
-      auto* thread_local_scoped_config =
-          const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
-      thread_local_scoped_config->removeRoutingScope(scoped_route_name);
-      return config;
-    });
+  ConfigSubscriptionCommonBase::onConfigUpdate();
+  if (any_applied) {
+    setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
   }
-
-  DeltaConfigSubscriptionInstance::onConfigUpdate();
-  setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
   stats_.config_reload_.inc();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s): {}",
+                                     StringUtil::join(exception_msgs, ", ")));
+  }
 }
 
+// TODO(stevenzzzz): see issue #7508, consider generalizing this function as it overlaps with
+// CdsApiImpl::onConfigUpdate.
+void ScopedRdsConfigSubscription::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  absl::flat_hash_map<std::string, envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
+  absl::flat_hash_map<uint64_t, std::string> scope_name_by_key_hash;
+  for (const auto& resource_any : resources) {
+    // Throws (thus rejects all) on any error.
+    auto scoped_route = MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
+        resource_any, validation_visitor_);
+    MessageUtil::validate(scoped_route);
+    const std::string scope_name = scoped_route.name();
+    auto scope_config_inserted = scoped_routes.try_emplace(scope_name, std::move(scoped_route));
+    if (!scope_config_inserted.second) {
+      throw EnvoyException(
+          fmt::format("duplicate scoped route configuration '{}' found", scope_name));
+    }
+    const envoy::api::v2::ScopedRouteConfiguration& scoped_route_config =
+        scope_config_inserted.first->second;
+    uint64_t key_fingerprint = MessageUtil::hash(scoped_route_config.key());
+    if (!scope_name_by_key_hash.try_emplace(key_fingerprint, scope_name).second) {
```

Comment:
Why not use the string as a key rather than the fingerprint?

Reply:
the key idea is that this comparison maybe done multiple times at which point computing a hash   and store it as key will be faster than comparing two strings directly.